rlclientlib/federation/sender_joined_log_provider.cc

```diff
@@ -0,0 +1,244 @@
+  // Check endianness requirement of format and make sure we get that right here.
+  output.reserve(output.size() + sizeof(uint32_t));
+  output.insert(
+      std::end(output), reinterpret_cast<uint8_t*>(&data), reinterpret_cast<uint8_t*>(&data) + sizeof(uint32_t));
+}
+
+int emit_filemagic_message(std::vector<uint8_t>& output)
+{
+  constexpr uint32_t filemagic = 0x42465756;
+  constexpr uint32_t version = 1;
+  emit_uint32(output, filemagic);
+  emit_uint32(output, version);
+  return 0;
+}
+
+int emit_regular_message(std::vector<uint8_t>& output, flatbuffers::FlatBufferBuilder& fbb,
+    flatbuffers::Offset<messages::flatbuff::v2::JoinedPayload> joined_payload)
+{
+  constexpr uint32_t regular = 0xFFFFFFFF;
+  emit_uint32(output, regular);
+
+  fbb.Finish(joined_payload);
+  auto buffer = fbb.Release();
+
+  uint32_t size = buffer.size();
+  emit_uint32(output, size);
+
+  output.reserve(output.size() + size);
+  output.insert(std::end(output), buffer.data(), buffer.data() + size);
+
+  uint32_t padding_size = size % 8;
+  output.reserve(output.size() + padding_size);
+  output.insert(std::end(output), padding_size, 0);
+
+  return 0;
+}
+}  // namespace
+
+namespace reinforcement_learning
+{
+RL_ATTR(nodiscard)
+int sender_joined_log_provider::create(std::unique_ptr<sender_joined_log_provider>& output,
+    const utility::configuration& config, i_trace* trace_logger, api_status* status)
+{
+  if (config.get_int(name::PROTOCOL_VERSION, 999) != 2)
+  { RETURN_ERROR_LS(trace_logger, status, invalid_argument) << " protocol version 2 required"; }
+
+  std::string eud_duration = config.get(name::EUD_DURATION, "UNSET");
+  if (eud_duration == "UNSET") { RETURN_ERROR_ARG(trace_logger, status, invalid_argument, "eudduration must be set"); }
+
+  std::chrono::seconds eud_offset;
+  RETURN_IF_FAIL(parse_eud(eud_duration, eud_offset, status));
+
+  output = std::unique_ptr<sender_joined_log_provider>(new sender_joined_log_provider(eud_offset, trace_logger));
+  return error_code::success;
+}
+
+sender_joined_log_provider::sender_joined_log_provider(std::chrono::seconds eud_offset, i_trace* trace_logger)
+    : _eud_offset(eud_offset), _trace_logger(trace_logger)
+{
+}
+
+RL_ATTR(nodiscard)
+int sender_joined_log_provider::invoke_join(std::unique_ptr<VW::io::reader>& batch, api_status* status)
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+  std::vector<uint8_t> result;
+  const auto eud_cutoff = std::chrono::system_clock::now() - _eud_offset;
+  emit_filemagic_message(result);
+  for (auto it = _interactions.cbegin(); it != _interactions.cend();)
+  {
+    flatbuffers::FlatBufferBuilder fbb;
+    std::vector<flatbuffers::Offset<messages::flatbuff::v2::JoinedEvent>> events;
+
+    const auto& item = *it;
+    const auto interaction_timestamp = std::get<0>(it->first);
+    const auto interaction_time = interaction_timestamp.to_time_point();
+    if (interaction_time <= eud_cutoff)
+    {
+      const auto reward_cutoff = interaction_time + _eud_offset;
```

Comment:
Let's think if we need this. I believe existing joiner is treating EUD not as hard cutoff but as "we cannot guarantee for data after EUD to be joined" which is guaranteed by line 149 already.